generator/src/main/java/org/apache/kafka/message/RequestApiScope.java

```diff
@@ -0,0 +1,33 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.message;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+public enum RequestApiScope {
+    @JsonProperty("zkBroker")
+    ZK_BROKER,
+
+    @JsonProperty("broker")
+    BROKER,
+
+    @JsonProperty("controller")
+    CONTROLLER,
+
+    @JsonProperty("raft")
+    RAFT;
```

Comment:
If we're talking about scopes, wouldn't it be more intuitive if one of them was `CLIENT`? Since these definitions go into the protocol definition files that are used by other clients, this could be used to differentiate between protocol apis that clients need to support versus protocol apis that are used for inter broker and broker -> controller communication.

Reply:
I had something like that originally, but I found it simpler to have a 1-1 mapping from listener to scope. All "CLIENT" APIs have to be present on both the zk and raft brokers, but it alone doesn't give us a way to distinguish between the two. We could have "ZK_CLIENT" and "RAFT_CLIENT" as an alternative. Then I guess we would also still need "ZK_BROKER" and "RAFT_BROKER." In any case, then each listener might have multiple scopes, which is a little more complex.

Perhaps this would be clearer if we had a name tying this back the listener more directly. Maybe we should get rid of the "scope" term and just call it "listeners"?

```json
"listeners": ["broker", "controller"]
```

By the way, you probably saw it already, but the "raft" scope was added here just for convenience in `TestRaftServer`. I could use the "controller" scope instead and remove this.

Reply:
These are good points. I think if we keep the approach suggested here, `listeners` may be better. However, I am still not sure if it would not be better to have `ZK_CLIENT` and `RAFT_CLIENT` (or something along those lines). Then, as we drop support for ZK, it would work pretty nicely. Generally, I think the json files would ideally be our protocol definition versus something where we capture implementation details.

Reply:
@ijuma : unfortunately we don't have a clear separation between clients and brokers at the protocol level.  As you know, if a random node connects to the broker and asks for ApiVersions, the broker doesn't (yet) know if the other node is another broker or if it's a client.  So it wouldn't help to label APIs as CLIENT vs. BROKER, since the broker has to send back both anyway.

On the other hand, we can definitely put APIs into "zk broker", "kip-500 zk", and "kip500 controller" and use those buckets to figure out what to send in the ApiVersionsRequest.

@hachikuji : The "raft" scope seems a little weird since all the other scopes map to listeners on servers.  The TestRaftServer is just an internal junit thing, right?  So it's probably fine if it just uses the Controller scope and returns `UnsupportedVersionException` for things it doesn't support.  I don't think we should have a separate scope for this unless it somehow impacts the real servers.

Reply:
Ok, I will get rid of it.